rust/parsing.rs

```diff
@@ -0,0 +1,945 @@
+    pub second_datetime: Option<ParsedDateTime>,
+}
+
+impl Parsed {
+    pub fn new() -> Parsed {
+        Parsed {
+            datetime: None,
+            duration: None,
+            second_datetime: None,
+        }
+    }
+}
+
+pub struct Parser<'a> {
+    /// Input to parse.
+    src: &'a str,
+    /// Iterator used for getting characters from `src`.
+    chars: CharIndices<'a>,
+    /// Current byte offset into `src`.
+    idx: usize,
+    /// Current character
+    current: char,
+}
+
+impl<'a> Parser<'a> {
+    /// Creates a new parser from a &str.
+    pub fn new(input: &'a str) -> Parser<'a> {
+        let mut p = Parser {
+            src: input,
+            chars: input.char_indices(),
+            idx: 0,
+            current: '\0',
+        };
+        p.inc();
+        p
+    }
+
+    /// Increments the parser if the end of the input has not been reached.
+    /// Returns whether or not it was able to advance.
+    fn inc(&mut self) -> bool {
+        match self.chars.next() {
+            Some((i, ch)) => {
+                self.idx = i;
+                self.current = ch;
+                true
+            }
+            None => {
+                self.idx = self.src.len();
+                self.current = '\0';
+                false
+            }
+        }
+    }
+
+    fn parse_error(&mut self, message: String) -> ParseError {
+        ParseError {
+            index: self.idx,
+            message,
+        }
+    }
+
+    fn unexpected_character_error(
+        &mut self,
+        field_name: &str,
+        expected_character_count: usize,
+    ) -> ParseError {
+        if self.end() {
+            return self.parse_error(format!(
+                "Unexpected end of string while parsing {}. Expected {} more character{}.",
+                field_name,
+                expected_character_count,
+                if expected_character_count != 1 {
+                    "s"
+                } else {
+                    ""
+                }
+            ));
+        }
+
+        self.parse_error(format!(
+            "Invalid character while parsing {}: {}.",
+            field_name, self.current,
+        ))
+    }
+
+    /// Returns true if the parser has reached the end of the input.
+    fn end(&self) -> bool {
+        self.idx >= self.src.len()
+    }
+
+    fn parse_integer(&mut self, length: usize, field_name: &str) -> Result<u32, ParseError> {
+        let mut value: u32 = 0;
+
+        for i in 0..length {
+            if self.end() {
+                return Err(self.parse_error(format!(
+                    "Unexpected end of string while parsing \"{}\". Expected {} more character{}",
+                    field_name,
+                    length - i,
+                    if (length - i) != 1 { "s" } else { "" }
+                )));
+            }
+
+            if self.current >= '0' && self.current <= '9' {
+                value = 10 * value + self.current.to_digit(10).unwrap();
```

Comment:
This can imo be better expressed using `if-let` and `to_digit` directly. `.unwrap` is usually unidiomatic, even if it can't fail
```suggestion
            if let Some(digit) = self.current.to_digit(10) {
                value = 10 * value + digit;
```
also on lines 480 and 530